client/src/util/stateManager/world.js

```diff
@@ -37,78 +46,186 @@ Notable keys in the world object:
 * varData: a cache of expression columns, stored in a Dataframe.  Cache
   managed by controls reducer.
 
+* unclipped: will contain unclipped variants of all potentiall clipped
+  dataframes (obsAnnotations, varData).
+
 */
 
 function templateWorld() {
+  const obsAnnotations = Dataframe.Dataframe.empty();
+  const varAnnotations = Dataframe.Dataframe.empty();
+  const obsLayout = Dataframe.Dataframe.empty();
+  const varData = Dataframe.Dataframe.empty(null, new Dataframe.KeyIndex());
   return {
     /* schema/version related */
     schema: null,
     nObs: 0,
     nVar: 0,
+    clipQuantiles: { min: 0, max: 1 },
 
     /* annotations */
-    obsAnnotations: Dataframe.Dataframe.empty(),
-    varAnnotations: Dataframe.Dataframe.empty(),
+    obsAnnotations,
+    varAnnotations,
 
     /* layout of graph. Dataframe. */
-    obsLayout: Dataframe.Dataframe.empty(),
+    obsLayout,
 
-    /*
-    Var data columns - subset of all data (may be empty)
-    */
-    varData: Dataframe.Dataframe.empty(null, new Dataframe.KeyIndex())
+    /* Var data columns - subset of all data (may be empty) */
+    varData,
+
+    /* unclipped dataframes - subset, but not value clipped */
+    unclipped: {
+      obsAnnotations,
+      varData
+    }
   };
 }
 
-export function createWorldFromEntireUniverse(universe) {
-  const world = templateWorld();
-
+function clipDataframe(
+  df,
+  lowerQuantile,
+  upperQuantile,
+  quantileF,
+  clipPredicate = () => true,
+  value = Number.NaN
+) {
   /*
-  public interface follows
+  For all columns in the dataframe, clip all values above or below specified 
+  quantiles to `value`, IFF clipPredicate returns true.
+
+  Returns a clipped copy - does not mutate original.
+
+  clipPredicate must have signature: (dataframe, colIndex, colLabel) => boolean
+
+  quantileF must have signature:  (label, qval) => number
   */
+  if (lowerQuantile < 0) lowerQuantile = 0;
+  if (upperQuantile > 1) upperQuantile = 1;
+  if (lowerQuantile === 0 && upperQuantile === 1) return df;
+
+  const keys = df.colIndex.keys();
+  return df.mapColumns((col, colIdx) => {
+    const colLabel = keys[colIdx];
+    if (!clipPredicate(df, colIdx, colLabel)) return col;
```

Comment:
It is documented in the comment describing the function arguments (a few lines before the code in question).    Would it help if I duplicate the comment (or was it confusing)?

```
/*
  For all columns in the dataframe, clip all values above or below specified 
  quantiles to "value", IFF clipPredicate returns true.

  Returns a clipped copy - does not mutate original.

  clipPredicate must have signature: (dataframe, colIndex, colLabel) => boolean

  quantileF must have signature:  (label, qval) => number
  */
```

Reply:
The `IFF clipPredicate returns true` implied that that was the "should I clip" test, but I also thought it could be something else or something broader that was being tested

Reply:
clarifying the role of clipPredicate in the above comment would be all I would need

Reply:
it is literally a "should I clip" test.   The first sentence says "it will clip IIF clipPredicate returns true."

I suspect the confusion is that it is applied per-column, not per-Dataframe?  I'll clarify that.